apps/passport-server/migrations/58_password_reset_limit.sql

```diff
@@ -0,0 +1,98 @@
+-- in the Software without restriction, including without limitation the rights
+-- to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+-- copies of the Software, and to permit persons to whom the Software is
+-- furnished to do so, subject to the following conditions:
+
+-- The above copyright notice and this permission notice shall be included in all
+-- copies or substantial portions of the Software.
+
+-- THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+-- IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+-- FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+-- AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+-- LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+-- OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+-- SOFTWARE.
+
+-- This implements a "token bucket" approach to rate limiting, where each
+-- request causes a counter to decrease, where that counter is also re-filled
+-- based on time elapsed since the last refill. This allows us to ensure that
+-- requests are limited to a certain rate, without hard cut-offs around expiry
+-- or without needing to track each individual request.
+--
+-- Based on code here: https://github.com/fafl/token-bucket-postgres
+
+DROP FUNCTION IF EXISTS take_token(VARCHAR, VARCHAR);
+CREATE OR REPLACE FUNCTION take_token (event_type VARCHAR(100), event_id VARCHAR(100)) RETURNS boolean AS $$
+DECLARE
+    rate INTEGER;
+    tokens INTEGER;
+    extra_tokens INTEGER;
+    new_tokens INTEGER;
+    last_refill TIMESTAMP;
+    this_refill TIMESTAMP;
+BEGIN
+    -- Check if this user exists
+    SELECT hourly_limit INTO rate FROM rate_limit_types r WHERE r.event_type = $1;
+    IF rate IS NULL THEN
+        raise notice 'Event type % does not have a rate configured', $1;
+        RETURN FALSE;
+    END IF;
+
+    -- Lock the buckets until end of transaction
+    LOCK TABLE rate_limit_buckets IN EXCLUSIVE MODE;
+
+    -- Read current tokens and last take
+    SELECT b.remaining, b.last_refill INTO tokens, last_refill FROM rate_limit_buckets b WHERE b.event_type = $1 AND b.event_id = $2;
+    IF tokens IS NULL THEN
+        tokens := rate; -- Start with the max amount of tokens
+        last_refill = now();
+        raise notice 'Setting up a bucket for type % and id % with % tokens', $1, $2, tokens;
+        INSERT INTO rate_limit_buckets VALUES ($1, $2, tokens, last_refill);
+    END IF;
+
+    -- Calculate newly generated tokens since last call
+    extra_tokens := floor(
+        EXTRACT(EPOCH FROM (now() - last_refill) * rate / 3600.0)
+    )::int;
+    this_refill := last_refill + (extra_tokens * interval '1 second' * 3600.0 / rate);
+    new_tokens := LEAST(rate, tokens + extra_tokens);
+    raise notice 'Bucket %:% has % tokens, last batch generated at %', $1, $2, new_tokens, this_refill;
```

Comment:
question: do these notices get propagated up to the db client in typescript? if not, would that be possible somehow?

Reply:
No. I've removed the notices and converted the error notice to an exception. There's an example of this being caught at the end of the rate limit test file.